### PR TITLE
8338884: java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#tmp fails on alinux3

### PR DESCRIPTION
**Pull request by sendaoYan**
Hi all,
On alinux3(alibaba cloud linux version 3) system, the `/tmp` disk partition is mounted as tmpfs filesystem type, this filesystem type doesn't support create time(birth time).

Before this PR, this test [check](https://github.com/openjdk/jdk/blob/master/test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#L110) if there is `statx` system call present or not to determise the test environment support birth time or not. I think it's not enough when the tested filesystem type is `tmpfs`. When the tested filesystem type is `tmpfs`, then the tested file doesn't support birth time.

On RHEL 8 tmpfs doesn't seem to support birth time, but on F39 tmpfs does seem to support birth time. Looks like this might be related to the kernel version. It's difficult to enumerate all the combination of file system type and linux kernel version to determine the testd file support birth time or not. So in this PR, I get the result from `statx` linux syscall, to determine the testd file support birth time or not.

Test fix only, the change has been verified, risk is low.

Additional test:

- [x] Alinux3 glibc:2.32
1.  /tmp/file supportsCreationTimeRead == false
2. ./file supportsCreationTimeRead == true
- [x] CentOS7 docker container glibc:2.17
1.  /tmp/file supportsCreationTimeRead == false
2. ./file supportsCreationTimeRead == false

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8338884](https://bugs.openjdk.org/browse/JDK-8338884): java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#tmp fails on alinux3 (**Bug** - P3)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20687/head:pull/20687` \
`$ git checkout pull/20687`

Update a local copy of the PR: \
`$ git checkout pull/20687` \
`$ git pull https://git.openjdk.org/jdk.git pull/20687/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20687`

View PR using the GUI difftool: \
`$ git pr show -t 20687`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20687.diff">https://git.openjdk.org/jdk/pull/20687.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20687#issuecomment-2306140521)

### CODE REVIEW

**Comment by bplb** [comment]
What happens in the test without this change if the fix for [JDK-8338696](https://bugs.openjdk.org/browse/JDK-8338696) is applied?

**Comment by sendaoYan** [comment]
As descript in [JDK-8338884](https://bugs.openjdk.org/browse/JDK-8338884), this test will report fails `java.lang.RuntimeException: Creation time should not have changed`.
On alinux3(alibaba cloud system v3) the /tmp disk partition is mounted as tmpfs filesystem, and linux doesn't support birth time with this type filesystem. 
Without this change before the fix [JDK-8338696](https://bugs.openjdk.org/browse/JDK-8338696) is applied, `then the creation time is set to the epoch (1970-01-01T00:00:00Z)` so this test will [throw](https://github.com/openjdk/jdk/blob/master/test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#L84) `SkippedException`. 
Without this change after the fix [JDK-8338696](https://bugs.openjdk.org/browse/JDK-8338696) is applied, the `SkippedException` do not throw any more, and the `supportsCreationTimeRead` set to true, actually on alinux3 the file in /tmp partition doesn't support create time(birth time), and create time always equals to modified time, then this test report [fails](https://github.com/openjdk/jdk/blob/master/test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#L126).

**Comment by bplb** [comment]
> Without this change after the fix [JDK-8338696](https://bugs.openjdk.org/browse/JDK-8338696) is applied, the `SkippedException` do not throw any more, and the `supportsCreationTimeRead` set to true, actually on alinux3 the file in /tmp partition doesn't support create time(birth time), and create time always equals to modified time, then this test report [fails](https://github.com/openjdk/jdk/blob/master/test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#L126).

So you are saying that with the fix for JDK-8338696 applied, this test fails with `RuntimeException`?

**Comment by sendaoYan** [comment]
> So you are saying that with the fix for JDK-8338696 applied, this test fails with `RuntimeException`?

Yes, without this change, with the fix for JDK-8338696 applied, this test fails with `[RuntimeException](https://github.com/openjdk/jdk/blob/master/test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#L126)`

**Comment by sendaoYan** [comment]
To reproduce the fails, test command on most linux system:
```shell
rm -rf /dev/shm/tmp/ ; jtreg -v:fail,error -w /dev/shm/tmp -nr test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java#cwd
```

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 1]
```diff
@@ -108,6 +108,10 @@ static void test(Path top) throws IOException {
         } else if (Platform.isLinux()) {
             // Creation time read depends on statx system call support
             supportsCreationTimeRead = Linker.nativeLinker().defaultLookup().find("statx").isPresent();
+            // Linux system doesn't support birth time on tmpfs filesystem for now
+            if( Files.getFileStore(file).type().contentEquals("tmpfs") ) {
```
Comment (bplb) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:112]: `if( ` should be `if (` but I would suggest instead:

```
// Creation time read depends on statx system call support and on the file
// system storing the birth time. The tmpfs file system type does not store
// the birth time.
boolean statxIsPresent = Linker.nativeLinker().defaultLookup().find("statx").isPresent();
if (statxIsPresent && !Files.getFileStore(file).type().contentEquals("tmpfs")) {
    supportsCreationTimeRead = true;
}
```

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: Thanks for your review and suggestion. The code has been updated accroding your suggestion.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 2]
```diff
@@ -106,8 +106,13 @@ static void test(Path top) throws IOException {
...
         } else if (Platform.isLinux()) {
-            // Creation time read depends on statx system call support
-            supportsCreationTimeRead = Linker.nativeLinker().defaultLookup().find("statx").isPresent();
+            // Creation time read depends on statx system call support and on the file
+            // system storing the birth time. The tmpfs file system type does not store
+            // the birth time.
+            boolean statxIsPresent = Linker.nativeLinker().defaultLookup().find("statx").isPresent();
+            if (statxIsPresent && !Files.getFileStore(file).type().contentEquals("tmpfs")) {
+                supportsCreationTimeRead = true;
+            }
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:115]: Noting here that the test could still fail on some other filesystem which might not support birth time. For example `nfs` on Linux (IIRC).

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:115]: It is difficult to enumerate which file types in Linux do and do not support creation time.
How about add a function that use `stat -c` linux command line to determise support or not.

+    /**
+     * read the output of linux command `stat -c "%w" file`, if the output is "-",
+     * then the file system doesn't support birth time
+     */
+    public static boolean supportBirthTimeOnLinux(Path file) {
+        try {
+            String filePath = file.toAbsolutePath().toString();
+            ProcessBuilder pb = new ProcessBuilder("stat", "-c", "%w", filePath);
+            pb.redirectErrorStream(true);
+            Process p = pb.start();
+            BufferedReader b = new BufferedReader(new InputStreamReader(p.getInputStream()));
+            String l = b.readLine();
+            if (l != null && l.equals("-")) { return false; }
+        } catch(Exception e) {
+        }
+        return true;
+    }

Comment (jerboaa) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:115]: It's a possibility. I'm fine with the current "tmpfs" special case in the test, but I guess what you proposed above would be more robust. So if others are OK with such an approach, that would be fine with me as well.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 3]
```diff
@@ -141,6 +148,25 @@ static void test(Path top) throws IOException {
...
+        try {
+            String filePath = file.toAbsolutePath().toString();
+            ProcessBuilder pb = new ProcessBuilder("stat", "-c", "%w", filePath);
+            pb.redirectErrorStream(true);
+            Process p = pb.start();
+            BufferedReader b = new BufferedReader(new InputStreamReader(p.getInputStream()));
+            String l = b.readLine();
+            if (l != null && l.equals("-")) { return false; }
+        } catch(Exception e) {
+        }
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:166]: This has several bugs, there is test is test infrastructure to do this correctly. My preference would be not include a launch of "stat" if possible.

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: The code of `launch of "stat"` has been reverted.

**Comment by sendaoYan** [comment]
`nfs` on linux also doesn't support create time:
```shell
> mount | grep "/vmfarm "
pepsi.jvm.alibaba.net:/vmfarm on /vmfarm type nfs4 (rw,relatime,vers=4.1,rsize=1048576,wsize=1048576,namlen=255,hard,proto=tcp,port=0,timeo=600,retrans=2,sec=sys,clientaddr=11.165.41.2,local_lock=none,addr=11.158.137.189)
> stat /vmfarm/mountVmFarm.sh 
  File: ‘/vmfarm/mountVmFarm.sh’
  Size: 71              Blocks: 8          IO Block: 1048576 regular file
Device: 3ch/60d Inode: 12          Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2024-02-28 17:03:54.713251462 +0800
Modify: 2020-03-03 11:56:41.548038922 +0800
Change: 2020-03-03 11:56:41.548038922 +0800
 Birth: -
```

**Comment by sendaoYan** [comment]
On RHEL 8 tmpfs doesn't seem to support birth time, but on F39 tmpfs does seem to support birth time. Looks like this might be related to the kernel version. It's difficult to enumerate all the combination of file system type and linux kernel version to determine the testd file support birth time or not. So in this PR, I get the output from `stat -c` linux command, to determine the testd file support birth ot not.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 4]
```diff
@@ -141,6 +148,24 @@ static void test(Path top) throws IOException {
...
+    public static boolean supportBirthTimeOnLinux(Path file) {
+        try {
+            String filePath = file.toAbsolutePath().toString();
+            ProcessBuilder pb = new ProcessBuilder("stat", "-c", "%w", filePath);
+            pb.redirectErrorStream(true);
+            Process p = pb.start();
+            BufferedReader b = new BufferedReader(new InputStreamReader(p.getInputStream()));
+            String l = b.readLine();
+            if (l != null && l.equals("-")) { return false; }
+        } catch(Exception e) {
```
Comment (AlanBateman) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:164]: There are several bugs/issues with this method, did you mean to bring it back? If we are launching subprocesses in this test then it can use ProcessTools infrastructure.

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:164]: It's difficult to enumerate all the combination of file system type and linux kernel version to determine the testd file support birth time or not, So I want to bring it back.
Thanks for the advice, I will use `ProcessTools infrastructure` later.

Comment (jerboaa) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:164]: An alternative could be to use FFM API to query the statx bufs mask the kernel returns testing for the birth time bit like the actual code does.

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:164]: I think this PR needed to backport to jdk21u and jdk17u in future. Maybe in jdk17u can't use`FFM API`?

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:164]: Maybe we can use this solution https://github.com/openjdk/jdk21u/pull/282

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+#endif
+
+// static native boolean linuxIsCreationTimeSupported()
+JNIEXPORT jboolean JNICALL
+Java_CreationTimeHelper_linuxIsCreationTimeSupported(JNIEnv *env, jclass cls)
+{
+#if defined(__linux__)
+    void* statx_func = dlsym(RTLD_DEFAULT, "statx");
+    return statx_func != NULL ? JNI_TRUE : JNI_FALSE;
+#else
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:35]: This is equivalent to the previous `Linker.nativeLinker().defaultLookup().find("statx").isPresent();` so isn't enough. You'd have to pass in the path and do an actual `statx` call, look at the returned statx buffer, get the mask and look whether birth time has been filled in.

See: https://github.com/openjdk/jdk/commit/c89a1c35bda9002ee687b3fa267f3ef9cba78b00

Only then it makes sense to go the jni route. Having said all that, for JDK head using FFM is preferred. Let's not consider backports just yet.

**Review by jerboaa — CHANGES_REQUESTED** [review]

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 5 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:35]: Sorry, I already realized that the current solution is incomplete, so I do not click the previous `Resolve conversation`. I will finish the solution according your advice these days, Thanks.

**Comment by sendaoYan** [comment]
I think this PR should be set as draft status, sorry for that.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java (hunk 4 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: The solution has been change to, get the result from `statx` linux syscall througt `FFM API`, to determine the testd file support birth ot not.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 5 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c]: The solution has been change to, get the result from `statx` linux syscall througt `FFM API`, to determine the testd file support birth ot not.

**Comment by sendaoYan** [comment]
Sorry, build fails on macos and windows, I will fix it right now.

**Comment by bplb** [comment]
Line 84 of the test: "birth time not support for: " should be changed to "birth time not supported for: ".

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+ * questions.
+ */
+#include "export.h"
+#if (__STDC_VERSION__ >= 199901L)
+  #include <stdbool.h>
+#else
+  #define bool int
+  #define true 1
+  #define false 0
+#endif
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:30]: Do we really need this? Since https://openjdk.org/jeps/347 test libraries should be compiled with `--std=c11`.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    ret = my_statx_func(AT_FDCWD, file, atflag, mask, &stx);
+
+    #ifdef DEBUG
+    printf("birth time = %ld\n", stx.stx_btime.tv_sec);
+    #endif
+    if (ret != 0) {
+        return false;
+    }
+    if (stx.stx_mask & STATX_BTIME)
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:121]: `if (stx.stx_mask & STAX_BTIME) != 0)` please. Also, please add a comment mentioning that "on some systems where `statx` is available birth time might still not be supported as it's file system specific. The only reliable way to check for support is looking at the filled in STATX_BTIME bit in the returned statx buffer mask."

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+#include <dlfcn.h>
+#ifndef STATX_BASIC_STATS
+#define STATX_BASIC_STATS 0x000007ffU
+#endif
+#ifndef STATX_BTIME
+#define STATX_BTIME 0x00000800U
+#endif
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:46]: Why is this needed?

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 9]
```diff
@@ -81,7 +81,7 @@ static void test(Path top) throws IOException {
             System.out.println("creationTime.toMillis() == " + creationTime.toMillis());
             // If the file system doesn't support birth time, then skip this test
             if (creationTime.toMillis() == 0) {
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:83]: Is this still useful? After https://bugs.openjdk.org/browse/JDK-8338696 we'd never get the epoch (`0`). If birth time is not supported, the last modified time is being returned. I think this can go and the `SkippedException` branch removed as well.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 6 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:30]: I am not quite sure, I copied these codes from `src/java.base/unix/native/libjsig/jsig.c`. I will check it.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 8 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c]: That seems not needed indeed. I will delete that.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+            FunctionDescriptor.of(C_BOOL, C_POINTER));
+
+    // Helper so as to determine 'statx' support on the runtime system
+    // static boolean linuxIsCreationTimeSupported(String file);
+    static boolean linuxIsCreationTimeSupported(String file) throws Throwable {
+        try (var arena = Arena.ofConfined()) {
+            MemorySegment s = arena.allocateFrom(file);
+            return (boolean)methodHandle.invokeExact(s);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java]: Okey, code `return false early if statx syscall is not available` has been added.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 7 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c]: Thanks for your patient code review.
`STATX_BTIME` return statx buffer mask convert to boolean explicitly, and add comment to explain why get `STATX_BTIME` bit.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java (hunk 9 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: Thanks for your correction, the `SkippedException` for epoch (0) has been deleted.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 6 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:30]: I think I need some time to study that.

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:30]: I found other tests such as `test/jdk/java/foreign/libIntrinsics.c` only `#include <stdbool.h>` to use bool type. So I think these codes can change to only `#include <stdbool.h>`.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+    final static Linker abi = Linker.nativeLinker();
+    static final SymbolLookup lookup = SymbolLookup.loaderLookup();
+    final static MethodHandle methodHandle = abi.
+            downcallHandle(lookup.findOrThrow("linuxIsCreationTimeSupported"),
+            FunctionDescriptor.of(C_BOOL, C_POINTER));
+
+    // Helper so as to determine 'statx' support on the runtime system
+    // static boolean linuxIsCreationTimeSupported(String file);
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java:45]: This comment is a bit misleading.

```suggestion
    // Helper so as to determine birth time support on Linux.
    // Support is determined in a two-step process:
    // 1. Determine if `statx` system call is available. If available proceed, otherwise
    //    return false.
    // 2. Perform an actual `statx` call on the given file and check for birth time support
    //    in the mask returned from the call. This is needed, since some file systems, like
    //    nfs, don't support birth time even though the `statx` system call is available.
```

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    static boolean linuxIsCreationTimeSupported(String file) throws Throwable {
+        if (!abi.defaultLookup().find("statx").isPresent())
+            return false;
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java:48]: I personally prefer using `if (...) { /* do something */ }` (add curly brackets).

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+static statx_func* my_statx_func = NULL;
+#endif  //#defined(__linux__)
+
+// static boolean linuxIsCreationTimeSupported(char* file)
+EXPORT bool linuxIsCreationTimeSupported(char* file) {
+#if defined(__linux__)
+    struct my_statx stx;
+    int ret, atflag = AT_SYMLINK_NOFOLLOW;
+    memset(&stx, 0xbf, sizeof(stx));
```
Comment (jerboaa) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:91]: Why `0xbf`?

How about `struct my_statx stx = {0}`?

**Review by jerboaa — COMMENTED** [review]
Seems mostly fine now.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java (hunk 12 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTimeHelper.java]: Thanks for your patient code review again.
1. curly brackets in function linuxIsCreationTimeSupported has been added; 2. comment for function linuxIsCreationTimeSupported has been complemented.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c (hunk 13 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:91]: I implement this code imitate from [test-statx.c](https://github.com/torvalds/linux/blob/master/samples/vfs/test-statx.c#L251), and I'm not quite sure why initial to `0xbf`. I think initial to zero will be Okey.

**Comment by sendaoYan** [comment]
Because this test invoke native libary, so mark this test as `/native`.

**Review by jerboaa — APPROVED** [review]
This looks good to me.

**Comment by bplb** [comment]
For whatever it shows, the test passed on Linux, macOS, and Windows in our CI.

**Comment by jerboaa** [comment]
> For whatever it shows, the test passed on Linux, macOS, and Windows in our CI.

@bplb Do you mean the updated one in this PR or the old one without this patch?

**Comment by bplb** [comment]
> > For whatever it shows, the test passed on Linux, macOS, and Windows in our CI.
> 
> @bplb Do you mean the updated one in this PR or the old one without this patch?

@jerboaa The latest version in this PR.

**Comment by jerboaa** [comment]
> > > For whatever it shows, the test passed on Linux, macOS, and Windows in our CI.
> > 
> > 
> > @bplb Do you mean the updated one in this PR or the old one without this patch?
> 
> @jerboaa The latest version in this PR.

OK, thanks!

**Comment by sendaoYan** [comment]
> > > For whatever it shows, the test passed on Linux, macOS, and Windows in our CI.
> @jerboaa The latest version in this PR.

Thanks for the test.

**Comment by sendaoYan** [comment]
@AlanBateman Can you review this again, thanks.

**Comment by AlanBateman** [comment]
> @AlanBateman Can you review this again, thanks.

I don't have cycles right now but I'm sure @bplb can help.

**Comment by bplb** [comment]
> I don't have cycles right now but I'm sure @bplb can help.

It's now in my queue.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 14]
```diff
@@ -122,8 +121,11 @@ static void test(Path top) throws IOException {
             Instant plusHour = Instant.now().plusSeconds(60L * 60L);
             Files.setLastModifiedTime(file, FileTime.from(plusHour));
             FileTime current = creationTime(file);
-            if (!current.equals(creationTime))
+            if (!current.equals(creationTime)) {
+                System.out.println("current = " + current);
```
Comment (bplb) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:125]: I think it would be better to use `System.err` on these two lines, as was done at line 81.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    if (file == NULL) {
+        printf("input file error!\n");
+        return false;
+    }
+
+    ret = my_statx_func(AT_FDCWD, file, atflag, mask, &stx);
+
+    #ifdef DEBUG
```
Comment (bplb) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/libCreationTimeHelper.c:105]: This `DEBUG` block could be removed.

**Comment by bplb** [comment]
The test passed on Linux, macOS, and Windows in our internal test platform. I think after the two most recent points mentioned above are addressed it is good to go.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java (hunk 14 above)
Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: Thanks for the review. These two `System.out` has been replaced as `System.err`.

**Comment by sendaoYan** [comment]
> The test passed on Linux, macOS, and Windows in our internal test platform. I think after the two most recent points mentioned above are addressed it is good to go.

Thanks for the verify and review. The two advices has been addressed.

File: test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java [hunk 16]
```diff
@@ -79,13 +78,8 @@ static void test(Path top) throws IOException {
         Instant now = Instant.now();
         if (Math.abs(creationTime.toMillis()-now.toEpochMilli()) > 10000L) {
             System.out.println("creationTime.toMillis() == " + creationTime.toMillis());
```
Comment (bplb) [inline test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java:80]: Sorry but would you please change line 80 to use `System.err` also? Thanks.

Comment (sendaoYan) [reply test/jdk/java/nio/file/attribute/BasicFileAttributeView/CreationTime.java]: Sorry for miss the replaced. The line 80 has been changed.

**Review by bplb — APPROVED** [review]
Looks all right now.

**Comment by sendaoYan** [comment]
Thanks all for the review.
/integrate

**Review by jerboaa — APPROVED** [review]

**Comment by sendaoYan** [comment]
Thanks all for the review.
/integrate
